src/VisualStudio/Core/Def/Experimentation/VisualStudioExperimentationService.cs

```diff
@@ -2,58 +2,55 @@
 
 using System;
 using System.Composition;
-using System.Reflection;
-using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using System.Threading;
+using System.Threading.Tasks;
 using Microsoft.CodeAnalysis.Experiments;
 using Microsoft.CodeAnalysis.Host.Mef;
 using Microsoft.Internal.VisualStudio.Shell.Interop;
 using Microsoft.VisualStudio.Shell;
+using Telemetry = Microsoft.VisualStudio.Experimentation;
 
 namespace Microsoft.VisualStudio.LanguageServices.Experimentation
 {
     [Export(typeof(VisualStudioExperimentationService))]
-    [ExportWorkspaceService(typeof(IExperimentationService), ServiceLayer.Host), Shared]
-    internal class VisualStudioExperimentationService : ForegroundThreadAffinitizedObject, IExperimentationService
+    [ExportWorkspaceService(typeof(IExperimentationServiceFactory), ServiceLayer.Host), Shared]
+    internal class VisualStudioExperimentationService : IExperimentationServiceFactory
     {
-        private readonly object _experimentationServiceOpt;
-        private readonly MethodInfo _isCachedFlightEnabledInfo;
-        private readonly IVsFeatureFlags _featureFlags;
+        private readonly IAsyncServiceProvider _serviceProvider;
+        private IExperimentationService _experimentationService;
 
         [ImportingConstructor]
         [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
-        public VisualStudioExperimentationService(IThreadingContext threadingContext, SVsServiceProvider serviceProvider)
-            : base(threadingContext)
+        public VisualStudioExperimentationService(SVsServiceProvider serviceProvider)
         {
-            object experimentationServiceOpt = null;
-            MethodInfo isCachedFlightEnabledInfo = null;
-            IVsFeatureFlags featureFlags = null;
+            _serviceProvider = (IAsyncServiceProvider)serviceProvider;
+        }
 
-            threadingContext.JoinableTaskFactory.Run(async () =>
+        public async ValueTask<IExperimentationService> GetExperimentationServiceAsync(CancellationToken cancellationToken)
+        {
```

Comment:
I am not sure why it moved to this factory model.

Reply:
The separation is described in the initial PR description above.

Reply:
The explanation doesn't make sense to me -- it says both the IExperimentationService can be fetched without threading dependencies (which is incorrect since it has UI thread dependencies -- see comments below), and then says the second part is without threading dependencies. I'm assuming that the description is wrong and you were trying to split this between a UI-thread dependent and UI-thread independent portion and just have a typo there.

Reply:
> ... it says both the IExperimentationServiceFactory can be fetched without threading dependencies (which is incorrect since it has UI thread dependencies - ...

It says `IExperimentationServiceFactory` can be fetched from the MEF catalog without threading dependencies. This statement is correct. The service cannot be _used_ without threading dependencies, but that's normal for asynchronous services.

> ... then says IExperimentationService can be used without threading dependencies ...

This is also correct. Once you have an instance of `IExperimentationService`, that instance can be used without threading dependencies.

> ... if both can be used without threading dependencies, why split them?

Only `IExperimentationService` can be used without threading dependencies.